### PR TITLE
[draft] MXFP8 quantized TitanRL

### PR DESCRIPTION
**Pull request by andrewor14**
TODO: update this description with more details

High-level so far:

- **MXFP8 fwd + BF16 bwd MoE.** The bottleneck for RL is typically in rollouts so we keep the backward in high precision for now for better training stability. We can explore quantizing the backward later if it provides meaningful perf gains. We only quantize MoE for now because initial microbenchmarks did not find significant gains for linears for models like Qwen3-30B-A3B, can add linear quantization later if needed.

- **Weight sync lifecycle owns the quantized weights**. Similar to https://github.com/pytorch/torchtitan/pull/4203, we only maintain 1 copy of mxfp8 weights as the source of truth (but for the generator, since there's no FSDP there). Trainer pre-quantizes weights to mxfp8 before passing them to generator (`quantize_expert_state_dict_to_mxfp8 `). Generator generates rollouts from pre-quantized weights directly (`MXFP8InferenceGroupedExperts`)

- **Adds padding functionality to DeepEP.** DeepEP is needed to enable cudagraphs on the generator to avoid the [unpinned D2H sync issue](https://github.com/pytorch/torchtitan/blob/7a401fbac96e4303bf4ba7c9c4def9a8b157eaa0/torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py#L634). However, TorchTitan's DeepEP dispatcher does not support padding yet, and our activation quantization kernel requires padding to 32 (or 128 for cutedsl), so we need to add that functionality.

- **TorchAO / TorchTitan split.** TorchTitan will maintain all workflows and only import kernels from TorchAO. For example, the autograd function `_MXFP8GroupedMMFwdBF16Bwd` is adapted from `_MXFP8GroupedMM` from TorchAO. TODO: we will likely replace this autograd function with a more FSDP-aware one following the MoE version of https://github.com/pytorch/torchtitan/pull/4203.

### CODE REVIEW

File: torchtitan/overrides/mxfp8_inference_grouped_experts.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+# MXFP8 scaling-group size, along the contracting dim.
+_BLOCK_SIZE = 32
+# Tiles the blocked scale layout pads to, mirroring torchao's
+# ``triton_mx_block_rearrange_per_group_3d``.
+_SWIZZLE_ROW_TILE = 128
+_SWIZZLE_COL_TILE = 4
+
+
+class MXFP8InferenceGroupedExperts(Module):
```
Comment (tianyu-l) [inline torchtitan/overrides/mxfp8_inference_grouped_experts.py:55]: Would like to work together and try to find a way to avoid defining inference-specific modules. For attention we had to do so because KV cache management in inference is handled by vLLM attention class. For quantization, it sounds less necessary.

File: torchtitan/distributed/deepep/deepep.py
Comment (tianyu-l) [inline torchtitan/distributed/deepep/deepep.py:1]: From what I understand, DeepEP won't be the most efficient backend on Blackwell. Right now we are looking at DistMoE, which is to be OSS'ed soon. So prefer to work over there together.
